### PR TITLE
Show error output on failed executable additions

### PR DESCRIPTION
**Pull request by pirkla**

### CODE REVIEW

File: PPPC Utility.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -83,11 +84,12 @@
 		6EC409F2214D8FFA00BE4F17 /* TCCProfileViewController.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = TCCProfileViewController.swift; sourceTree = "<group>"; };
 		6EC409F4214D95D200BE4F17 /* TCCProfile.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = TCCProfile.swift; sourceTree = "<group>"; };
 		6EC40A0F214DE3B200BE4F17 /* Executable.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = Executable.swift; sourceTree = "<group>"; };
-		6EC40A11214DF8FE00BE4F17 /* SecurityWrapper.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = SecurityWrapper.swift; sourceTree = "<group>"; };
+		6EC40A11214DF8FE00BE4F17 /* SecurityWrapper.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; name = SecurityWrapper.swift; path = "/Users/andrew.pirkl/xcodeGits/PPPC-Utility/Source/SecurityWrapper.swift"; sourceTree = "<absolute>"; };
```
Comment (macblazer) [inline PPPC Utility.xcodeproj/project.pbxproj:87]: This change should not be here.

File: Source/Model/LoadExecutableError.swift [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+extension LoadExecutableError: LocalizedError {
+    public var errorDescription: String? {
+        switch self {
+        case .identifierNotFound:
+            return "Bundle identifier could not be found."
+        case .resourceURLNotFound:
+            return "Resource URL could not be found."
+        case .codeRequirementError:
```
Comment (macblazer) [inline Source/Model/LoadExecutableError.swift:44]: This should pull in the underlying description (the associated string) and add that to the returned string.

```suggestion
        case .codeRequirementError(let underlyingDescription):
```

File: Source/View Controllers/TCCProfileViewController.swift [hunk 3]
```diff
@@ -314,10 +328,21 @@ class TCCProfileViewController: NSViewController {
...
         guard let openVC = segue.destinationController as? OpenViewController else { return }
+        guard let window = self.view.window else {
+            return
+        }
         if let button = sender as? NSButton, button == addAppleEventButton {
             Model.shared.current = executablesAC.selectedObjects.first as? Executable
             openVC.completionBlock = {
-                $0.forEach { self.insetIntoAppleEvents($0) }
+                $0.forEach {
+                    switch $0 {
```
Comment (macblazer) [inline Source/View Controllers/TCCProfileViewController.swift:338]: With two uses of $0 right next to each other, lets use a `result in` for the second one so this line becomes `switch result {`.

**Comment by macblazer** [comment]
This will close #37 when merged.

File: PPPC Utility.xcodeproj/project.pbxproj (hunk 1 above)
Comment (pirkla) [reply PPPC Utility.xcodeproj/project.pbxproj]: Woops! Reverted this change.

File: Source/Model/LoadExecutableError.swift (hunk 2 above)
Comment (pirkla) [reply Source/Model/LoadExecutableError.swift]: I went back and forth on this one a little bit, which is why I accidentally left the line to pass an associated value without using it.  I think in general more descriptive errors are better. It now looks like this:
Screen Shot 2020-05-07 at 7 15 44 AM

The problem is I don't think the underlying error is useful for general users, and for debugging that error is printed to the console. Thoughts?

File: Source/View Controllers/TCCProfileViewController.swift (hunk 3 above)
Comment (pirkla) [reply Source/View Controllers/TCCProfileViewController.swift]: Yep, fixed! I also switched the references to self to weak.

File: Source/Model/LoadExecutableError.swift (hunk 2 above)
Comment (macblazer) [reply Source/Model/LoadExecutableError.swift]: I like the more specific error message.

I think if general users have the error in the UI, they will be able to report a more complete issue here on GitHub.  If it's only in the console, general users will report a bug and our first step will always be to tell them "open the Console.app, try again in PPPC Utility, and then send us the actual underlying error message".

File: Source/View Controllers/TCCProfileViewController.swift [hunk 4]
```diff
@@ -150,10 +150,10 @@ class TCCProfileViewController: NSViewController {
         }
     }
 
-    fileprivate func showAlert(_ error: TCCProfileImportError, for window: NSWindow) {
+    fileprivate func showAlert<T: LocalizedError>(_ error: T, for window: NSWindow) {
```
[the remark this answers is not in the document]
Comment (macblazer) [reply Source/View Controllers/TCCProfileViewController.swift:153]: More info: generics do provide a better performance, so if this was being called in a tight loop I would say use generics.  But this is a function showing something in the UI so a few microsecond performance boost before the UI is displayed is not worth the added cognitive complexity of generics in this part of the code.

[the remark this answers is not in the document]
Comment (pirkla) [reply Source/View Controllers/TCCProfileViewController.swift]: Makes sense, and thanks for the insight! I switched it over, it definitely makes more sense for readability.

File: Source/View Controllers/TCCProfileViewController.swift [hunk 5]
```diff
@@ -351,17 +377,31 @@ extension TCCProfileViewController : NSTableViewDataSource {
...
-        
-        if tableView == executablesTable {
-            guard executablesAC.canInsert else { return false }
-            if shouldExecutableBeAdded(newExecutable) {
-                executablesAC.insert(newExecutable, atArrangedObjectIndex: row)
+        guard let window = self.view.window else { return false }
+        var canAdd = true
+        model.loadExecutable(url: url) {
+            [weak self] result in
+            switch result{
```
Comment (watkyn) [inline Source/View Controllers/TCCProfileViewController.swift:384]: some style related comments:
have a space before the opening brace here.

Line 388 the brace should be on the same line as 387. 

I added a card to put SwiftLint into the project so we can have an enforced style guile for the future:
https://github.com/jamf/PPPC-Utility/issues/55
So not a big deal right now, as that card will clean up any inconsistencies like this.

**Pull request merged**